### PR TITLE
Pulling Data from ARTAccess To UgandaEMR

### PR DESCRIPTION
**Pull request by dbaluku**
https://app.asana.com/0/1133167201254913/1200117590862827

How to Test This 
-Ensure you have some patients having same Clinic no like those from the API. A query for this can be found on the asana card
-Pull PR down to local and watch sync module
-set dhis2 uuid to wepUthscBkh 
-Reschedule the Receive ARTAccess Visit data Task to run in the shortest time possible 
-Check whether patient data has been synced

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java [hunk 1]
```diff
@@ -0,0 +1,494 @@
...
+            return;
+        }
+
+        if (!ugandaEMRHttpURLConnection.isConnectionAvailable()) {
+            return;
+        }
+
+
+
+        String ARTAccessServerUrlEndPoint="";
```
Comment (slubwama) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java:72]: Camel Case this Variable

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java [hunk 2]
```diff
@@ -0,0 +1,494 @@
...
+    protected final Log log = LogFactory.getLog(ReceiveVisitsDataFromARTAccessTask.class);
+
+    SyncGlobalProperties syncGlobalProperties = new SyncGlobalProperties();
+    UgandaEMRSyncService ugandaEMRSyncService = Context.getService(UgandaEMRSyncService.class);
+    ConceptService conceptService = Context.getConceptService();
+    VisitService visitService = Context.getVisitService();
+    EncounterService encounterService =Context.getEncounterService();
+    LocationService locationService = Context.getLocationService();
+    Location pharmacyLocation = locationService.getLocationByUuid("3ec8ff90-3ec1-408e-bf8c-22e4553d6e17");
+    EncounterType ARTCardEncounterType = encounterService.getEncounterTypeByUuid("8d5b2be0-c2cc-11de-8d13-0010c6dffd0f");
```
Comment (slubwama) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java:55]: Camel Case this Variable

**Review by slubwama — CHANGES_REQUESTED** [review]

File: api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java [hunk 3]
```diff
@@ -40,10 +40,7 @@
 import org.openmrs.module.ugandaemrsync.UgandaEMRSyncConfig;
 import org.openmrs.notification.Alert;
 
-import javax.net.ssl.HostnameVerifier;
-import javax.net.ssl.SSLContext;
-import javax.net.ssl.TrustManager;
-import javax.net.ssl.X509TrustManager;
+import javax.net.ssl.*;
```
[the remark this answers is not in the document]
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java (hunk 1 above)
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java]: Done

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java (hunk 2 above)
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java]: Done

**Review by slubwama — APPROVED** [review]

**Comment by slubwama** [comment]
@dbaluku resolve conflicts

File: api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java [hunk 4]
```diff
@@ -516,4 +512,28 @@ public static CloseableHttpClient createAcceptSelfSignedCertificateClient()
                 .setSSLSocketFactory(connectionFactory)
                 .build();
     }
+
+    public String getJson(String url) {
```
Comment (slubwama) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java:516]: @dbaluku why did you introduce this.

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java [hunk 5]
```diff
@@ -0,0 +1,494 @@
...
+
+            if (!ugandaEMRHttpURLConnection.isServerAvailable(artAccessServerUrlEndPoint)) {
+                log.error("server not available ");
+                return;
+            }
+
+        }
+
+        try {
+            results = ugandaEMRHttpURLConnection.getJson(artAccessServerUrlEndPoint);
```
Comment (slubwama) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java:86]: @dbaluku how are you taking care of authentication

Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java:86]: @slubwama the endpoint was open with no authentication

File: api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java (hunk 4 above)
Comment (dbaluku) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java:516]: Is there a similar function

File: api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java (hunk 5 above)
Comment (slubwama) [reply api/src/main/java/org/openmrs/module/ugandaemrsync/tasks/ReceiveVisitsDataFromARTAccessTask.java:86]: @dbaluku you always have to factor authentication in mind. Even when what you are testing with doesn't. 
We will be using HIM which requires either basic or token based.

File: api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java [hunk 6]
```diff
@@ -136,16 +136,13 @@ public StringBuffer getResponseString(BufferedReader bufferedReader) throws IOEx
...
-     * @param content
-     * @param facilityId
      * @param url
      * @param username
      * @param password
      * @return
      * @throws Exception
      */
-    public Map getByWithBasicAuth(String contentType, String content, String facilityId, String url, String username, String password, String resultType) throws Exception {
+    public Map getByWithBasicAuth(String url, String username, String password, String resultType) throws Exception {
```
Comment (slubwama) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java:145]: @dbaluku why are you changing the method signature. did you consider other possible use in and out of this module??

File: api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java [hunk 7]
```diff
@@ -170,6 +164,9 @@ public Map getByWithBasicAuth(String contentType, String content, String facilit
                 UsernamePasswordCredentials credentials = new UsernamePasswordCredentials(username, password);
 
                 httpGet.addHeader(new BasicScheme().authenticate(credentials, httpGet, null));
+                httpGet.addHeader("x-ugandaemr-facilityname", syncGlobalProperties.getGlobalProperty(GP_FACILITY_NAME));
```
Comment (slubwama) [inline api/src/main/java/org/openmrs/module/ugandaemrsync/api/UgandaEMRHttpURLConnection.java:167]: Make sure even without authentication the headers can be sent

**Comment by slubwama** [comment]
@dbaluku remove WIP from PR Name

**Pull request merged**